File: src/main/java/dev/zprestige/fire/module/visual/chams/Frame3DListener.java
package dev.zprestige.fire.module.visual.chams;

import dev.zprestige.fire.event.bus.EventListener;
import dev.zprestige.fire.event.impl.FrameEvent;
import dev.zprestige.fire.util.impl.RenderUtil;
import net.minecraft.client.Minecraft;
import net.minecraft.entity.Entity;

import java.awt.*;
import java.util.ArrayList;

public class Frame3DListener extends EventListener<FrameEvent.FrameEvent3D, Chams> {

    public Frame3DListener(final Chams chams) {
        super(FrameEvent.FrameEvent3D.class, chams);
    }

    @Override
    public void invoke(final Object object) {
        final FrameEvent.FrameEvent3D event = (FrameEvent.FrameEvent3D) object;
        final float partialTicks = event.getPartialTicks();
        final int fps = Minecraft.getDebugFPS();
        for (final Chams.PopEntity popEntity : new ArrayList<>(module.popEntities)) {
            final float alpha = popEntity.getAlpha();
            final Entity entity = popEntity.getEntity();
            if (module.animateVertical.GetSwitch()) {
                entity.posY += module.verticalAnimationSpeed.GetSlider() / 100.0f;
            }
            if (module.popFill.GetSwitch()) {
                final Color color = module.popFillColor.GetColor();
                module.prepareFill(new Color(color.getRed() / 255.0f, color.getGreen() / 255.0f, color.getBlue() / 255.0f, alpha / 255.0f));
                RenderUtil.renderEntity(entity, partialTicks);
                module.releaseFill();
            }
            if (module.popOutline.GetSwitch()) {
                final Color color = module.popOutlineColor.GetColor();
                module.prepareOutline(new Color(color.getRed() / 255.0f, color.getGreen() / 255.0f, color.getBlue() / 255.0f, alpha / 255.0f), module.popOutlineWidth.GetSlider());
                RenderUtil.renderEntity(entity, partialTicks);
                module.releaseOutline();
            }
            popEntity.updateAlpha(fps);
            if (popEntity.getAlpha() < 30) {
                module.popEntities.remove(popEntity);
                break;
            }
        }
    }
}

File: src/main/java/dev/zprestige/fire/module/visual/chams/Frame3DListener.kt
package dev.zprestige.fire.module.visual.chams

import dev.zprestige.fire.event.bus.EventListener
import dev.zprestige.fire.event.impl.FrameEvent.FrameEvent3D
import dev.zprestige.fire.util.impl.RenderUtil
import net.minecraft.client.Minecraft
import java.awt.Color

class Frame3DListener(chams: Chams) : EventListener<FrameEvent3D, Chams>(FrameEvent3D::class.java, chams) {
    
    override fun invoke(e: Any) {
        val event = e as FrameEvent3D
        val partialTicks = event.partialTicks
        val fps = Minecraft.getDebugFPS()
        for (popEntity in ArrayList(module.popEntities)) {
            val alpha = popEntity.alpha.toInt()
            val entity = popEntity.entity
            if (module.popAnimateVertical.GetSwitch()) {
                entity.posY += (module.popVerticalAnimationSpeed.GetSlider() / 100.0f).toDouble()
            }
            if (module.popFill.GetSwitch()) {
                val color = module.popFillColor.GetColor()
                module.prepareFill(Color(color.red, color.green, color.blue, alpha))
                RenderUtil.renderEntity(entity, partialTicks)
                module.releaseFill()
            }
            if (module.popOutline.GetSwitch()) {
                val color = module.popOutlineColor.GetColor()
                module.prepareOutline(Color(color.red, color.green, color.blue, alpha),
                    module.popOutlineWidth.GetSlider()
                )
                RenderUtil.renderEntity(entity, partialTicks)
                module.releaseOutline()
            }
            popEntity.updateAlpha(fps)
            if (popEntity.alpha < 30) {
                module.popEntities.remove(popEntity)
            }
        }
    }
}
